### PR TITLE
Fixes #39252 - Filter activation keys by environment_id

### PR DESCRIPTION
**Pull request by LadislavVasina1**
## Summary
- The `/katello/api/environments/:environment_id/activation_keys` endpoint
  was returning all activation keys in the organization instead of filtering
  by the specified lifecycle environment.
- The `environment_id` URL path parameter was accepted but never used in
  `index_relation` to filter results. Added the missing filter using the
  existing `with_environments` scope.
- Fixed typo in activation key fixture name ("Stating" → "Staging").
## Testing steps
- Have a Satellite and in a Default Organization create `Dev` environment with `Library` as prior LCE
- Create a CV=`Cv1` and publish a new version to both `Dev` and `Library` LCEs
- Create 1 AK `AK-Library-Test` with LCE=`Library` and CV=`Cv1`
- Create 1 AK `AK-Dev-Test` with LCE=`Dev` and CV=`Cv1`
Then check the API responses of the queries on AK from a specific Environments:
 ```
curl -s -u usrname:psswd 'https://localhost/katello/api/environments/<LCE_ID>/activation_keys' -k | python3 -m json.tool | grep -E '"(name|total|subtotal)"'
```
```
curl -s -u usrname:psswd 'https://localhost/katello/api/environments/<LCE_ID>/activation_keys' -k | python3 -m json.tool | grep -E '"(name|total|subtotal)"'
```

^^^ They should return only AK for the specific Environment, NOT all AK in the Organization
## Test plan
- [x] Added `test_index_filter_by_environment` to verify only activation
  keys belonging to the specified environment are returned
- [ ] Verify manually:
  - `curl -s -u usrname:psswd 'https://localhost/katello/api/environments/2/activation_keys' -k`
    returns only keys in environment 2
  - `curl -s -u usrname:psswd 'https://localhost/katello/api/environments/1/activation_keys' -k`
    returns only keys in environment 1
- [ ] Run full activation keys controller test suite:
  `ktest test/controllers/api/v2/activation_keys_controller_test.rb`

## Summary by CodeRabbit

* **New Features**
  * Activation keys API now supports filtering results by environment ID.

* **Bug Fixes**
  * Corrected a fixture name typo to ensure accurate test data.

* **Tests**
  * Added tests covering environment-based filtering, combined filters, empty results, and handling of nonexistent environment IDs.

### CODE REVIEW

**Comment by LadislavVasina1** [comment]
```
[katello]# ktest test/controllers/api/v2/activation_keys_controller_test.rb

Emptying /home/vagrant/foreman/jenkins/reports/unit
-- execute("SET CONSTRAINTS ALL DEFERRED;")
WARNING:  SET CONSTRAINTS can only be used in transaction blocks
   -> 0.0006s

# Running tests with run options --seed 10795:

.............................................Writing XML reports to /home/vagrant/foreman/jenkins/reports/unit


Finished tests in 8.287848s, 5.4296 tests/s, 25.5796 assertions/s.


45 tests, 212 assertions, 0 failures, 0 errors, 0 skips

Minitest Reporters: Mean Time Report (Samples: 7, Order: :avg :desc)
Avg: 3.789727161  Min: 1.574638968  Max: 9.465963075  Last: 8.256115156  Description: Katello::Api::V2::ActivationKeysControllerTest
Avg: 2.699367094  Min: 2.167069877  Max: 3.093556423  Last: 3.093556423  Description: Api::V2::HostsBulkActionsControllerTest::manage_notifications
```

**Review by jnagare-redhat — APPROVED** [review]
LGTM

File: app/controllers/katello/api/v2/activation_keys_controller.rb [hunk 1]
```diff
@@ -219,6 +219,7 @@ def index_relation
       activation_keys = activation_keys.where(:organization_id => @organization) if @organization
       activation_keys = activation_keys.with_content_view_environments(@content_view_environments) if @content_view_environments
       activation_keys = activation_keys.with_content_views(params[:content_view_id]) if params[:content_view_id]
+      activation_keys = activation_keys.with_environments(params[:environment_id]) if params[:environment_id]
```
Comment (lfu) [inline app/controllers/katello/api/v2/activation_keys_controller.rb]: Need to follow the pattern for `organization`, add the validation for `environment`.
```
before_action :find_optional_environment, :only => [:index] 

private

def find_optional_environment
  return unless params[:environment_id]
  @environment = KTEnvironment.readable.find(params[:environment_id])
end
```
Then update `index_relation`:
`activation_keys = activation_keys.with_environments(@environment.id) if @environment`

File: test/controllers/api/v2/activation_keys_controller_test.rb [hunk 2]
```diff
@@ -62,6 +62,17 @@ def test_index_filter_by_content_view
...
+      dev = katello_environments(:dev)
+      results = JSON.parse(get(:index, params: { :environment_id => dev.id }).body)
+
+      assert_response :success
+      assert_equal 1, results['results'].size
+      result_ids = results['results'].map { |r| r['id'] }
+      assert_includes result_ids, katello_activation_keys(:library_dev_staging_view_key).id
+      refute_includes result_ids, @activation_key.id
+    end
+
```
Comment (lfu) [inline test/controllers/api/v2/activation_keys_controller_test.rb:108]: Can we add edge case tests for the environment_id filtering?

File: app/controllers/katello/api/v2/activation_keys_controller.rb (hunk 1 above)
Comment (LadislavVasina1) [reply app/controllers/katello/api/v2/activation_keys_controller.rb]: ack, will do, thanks for review

File: app/controllers/katello/api/v2/activation_keys_controller.rb [hunk 3]
```diff
@@ -219,6 +220,7 @@ def index_relation
       activation_keys = activation_keys.where(:organization_id => @organization) if @organization
       activation_keys = activation_keys.with_content_view_environments(@content_view_environments) if @content_view_environments
       activation_keys = activation_keys.with_content_views(params[:content_view_id]) if params[:content_view_id]
+      activation_keys = activation_keys.with_environments(@environment.id) if @environment
       activation_keys = activation_keys.with_environments(params[:lifecycle_environments]) if params[:lifecycle_environments]
```
Comment (lfu) [inline app/controllers/katello/api/v2/activation_keys_controller.rb]: What happens if both `:environment_id` and `:lifecycle_environments` are specified with different values?  cc @jeremylenz 
Please add a test for this case.

Comment (jeremylenz) [reply app/controllers/katello/api/v2/activation_keys_controller.rb]: I think the last one will overwrite earlier ones. So `:lifecycle_environments` would win.

Comment (LadislavVasina1) [reply app/controllers/katello/api/v2/activation_keys_controller.rb]: @jeremylenz @lfu  Aren't the filters chained with `AND`? thus having different `:environment_id` and `:lifecycle_environments` would return 0

```
[devBox]# curl -s -u usr:psswd 'https://localhost/katello/api/environments/2/activation_keys?lifecycle_environments=1' -k | python3 -m json.tool | grep -E '"(name|total|subtotal)"'
    "total": 0,
    "subtotal": 0,
        "by": "name",
```

or am I missing something?

Comment (lfu) [reply app/controllers/katello/api/v2/activation_keys_controller.rb]: @LadislavVasina1 You are correct 💯 😄 
Thanks to Claude!
```
This doesn't actually override - both filters get applied (they're AND'd together in the SQL). Since the with_environments scope uses a WHERE clause on environment_id:

scope :with_environments, ->(lifecycle_environments) do
  joins(:content_view_environment_activation_keys => :content_view_environment).
    where("#{::Katello::ContentViewEnvironment.table_name}.environment_id" => lifecycle_environments)
end
Chaining them creates: WHERE environment_id = X AND environment_id IN (Y, Z), which would only return results if X is in the array [Y, Z].

To achieve true override behavior:

if params[:lifecycle_environments]
  activation_keys = activation_keys.with_environments(params[:lifecycle_environments])
elsif @environment
  activation_keys = activation_keys.with_environments(@environment.id)
end
```

Comment (LadislavVasina1) [reply app/controllers/katello/api/v2/activation_keys_controller.rb]: I still can add a test

Comment (lfu) [reply app/controllers/katello/api/v2/activation_keys_controller.rb]: If the API call is like:
```
https://localhost/katello/api/environments/<environment_id>/activation_keys
```
Maybe `environment_id` should have higher priority over `params[:lifecycle_environments]`? cc @jeremylenz

Comment (jeremylenz) [reply app/controllers/katello/api/v2/activation_keys_controller.rb]: That would be consistent with the logic from `def find_content_view_environments`.

Comment (jeremylenz) [reply app/controllers/katello/api/v2/activation_keys_controller.rb]: Actually, why are we not using that here :thinking:

File: test/controllers/api/v2/activation_keys_controller_test.rb [hunk 4]
```diff
@@ -62,6 +62,49 @@ def test_index_filter_by_content_view
...
+      assert_equal katello_activation_keys(:library_dev_staging_view_key).id, results['results'].first['id']
+    end
+
+    def test_index_filter_by_environment_id_and_lifecycle_environments_with_different_values
+      dev = katello_environments(:dev)
+      library = @organization.library
+      results = JSON.parse(get(:index, params: { :environment_id => dev.id, :lifecycle_environments => library.id }).body)
+
+      assert_response :success
+      assert_equal 0, results['results'].size
```
Comment (lfu) [inline test/controllers/api/v2/activation_keys_controller_test.rb]: We should get at least one result. From `:environment_id` I think.

Comment (LadislavVasina1) [reply test/controllers/api/v2/activation_keys_controller_test.rb]: In current implementation, we don't get that

File: app/controllers/katello/api/v2/activation_keys_controller.rb (hunk 3 above)
Comment (LadislavVasina1) [reply app/controllers/katello/api/v2/activation_keys_controller.rb]: @jeremylenz should I implement higher priority for `environment_id`?

Comment (jeremylenz) [reply app/controllers/katello/api/v2/activation_keys_controller.rb]: Seems fine, but as this is already an edge case - I don't expect these params to be incoherent frequently in real life - I don't consider it a blocking comment

Comment (LadislavVasina1) [reply app/controllers/katello/api/v2/activation_keys_controller.rb]: All right thanks.

> I don't expect these params to be incoherent frequently in real life

Agree on this

Comment (lfu) [reply app/controllers/katello/api/v2/activation_keys_controller.rb]: I would not rely on the behavior in real life to make it work. The code should work as expected.
How about this?
```
if @environment
  activation_keys = activation_keys.with_environments([@environment.id])
elsif params[:lifecycle_environments]
  activation_keys = activation_keys.with_environments(params[:lifecycle_environments])
end
```

Comment (LadislavVasina1) [reply app/controllers/katello/api/v2/activation_keys_controller.rb]: I can do that

**Review by lfu — APPROVED** [review]
LGTM 👍
